File: 17. Letter Combinations of a Phone Number/17. Letter Combinations of a Phone Number.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class Solution {

  public List<String> letterCombinations(String digits) {
    if (digits.equals("")) {
      return new ArrayList<>();
    }

    Map<Character, String> digitsMap = new HashMap<>();
    digitsMap.put('2', "abc");
    digitsMap.put('3', "def");
    digitsMap.put('4', "ghi");
    digitsMap.put('5', "jkl");
    digitsMap.put('6', "mno");
    digitsMap.put('7', "pqrs");
    digitsMap.put('8', "tuv");
    digitsMap.put('9', "wxyz");

    List<String> ans = new ArrayList<>();
    StringBuilder curr = new StringBuilder();

    backtrack(0, digits, digitsMap, ans, curr);
    return ans;
  }

  private void backtrack(
    int i,
    String digits,
    Map<Character, String> digitsMap,
    List<String> ans,
    StringBuilder curr
  ) {
    if (curr.length() == digits.length()) {
      ans.add(curr.toString());
      return;
    }

    String letters = digitsMap.get(digits.charAt(i));
    for (char c : letters.toCharArray()) {
      curr.append(c);
      backtrack(i + 1, digits, digitsMap, ans, curr);
      curr.deleteCharAt(curr.length() - 1);
    }
  }
}

File: 17. Letter Combinations of a Phone Number.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        digit_letter_mapping = {
            "2": "abc",
            "3": "def",
            "4": "ghi",
            "5": "jkl",
            "6": "mno",
            "7": "pqrs",
            "8": "tuv",
            "9": "wxyz"
        }

        ans = []
        curr = []

        def backtrack(index):
            if index == len(digits):
                ans.append("".join(curr[:]))
                return

            for char in digit_letter_mapping[digits[index]]:
                curr.append(char)
                backtrack(index + 1)
                curr.pop()

        backtrack(0)
        return ans
